exonum/src/events/noise/mod.rs

```diff
@@ -112,3 +114,174 @@ fn write_handshake_msg(
     let res = noise.write_handshake_msg();
     Box::new(done(res.map_err(|e| e.into())))
 }
+
+pub fn read_handshake_msg(
+    input: &[u8],
+    noise: &mut NoiseWrapper,
+) -> Box<Future<Item = (usize, Vec<u8>), Error = io::Error>> {
+    let res = noise.read_handshake_msg(input);
+    Box::new(done(res.map_err(|e| e.into())))
+}
+
+#[cfg(test)]
+mod tests {
+    use tokio_core::reactor::Core;
+    use tokio_core::net::{TcpListener, TcpStream};
+    use tokio_io::AsyncRead;
+    use futures::{done, Future, Stream};
+
+    use std::net::SocketAddr;
+    use std::thread;
+    use std::io;
+
+    use events::noise::wrapper::{NoiseWrapper, NOISE_MAX_MESSAGE_LENGTH,
+                                 NOISE_MIN_HANDSHAKE_MESSAGE_LENGTH};
+    use events::noise::{read, read_handshake_msg, write, HandshakeParams, HandshakeResult,
+                        NoiseHandshake};
+    use events::error::{into_other, log_error};
+    use events::codec::MessagesCodec;
+    use crypto::{gen_keypair_from_seed, Seed};
+    use std::time::Duration;
+
+    #[derive(Debug, PartialEq, Copy, Clone)]
+    pub enum HandshakeStep {
+        Default,
+        One(u8, usize),
```

Comment:
May be more verbose, `ClearText` \ `EncryptedText` variants would be better?
Or `BeforeHandshake` \ `AfterHandshake`.